nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/internal/P2ProxyIT.java

```diff
@@ -0,0 +1,192 @@
+
+  private static final String INVALID_PACKAGE_URL = PACKAGE_BASE_PATH + INVALID_PACKAGE_NAME;
+
+  private P2Client proxyClient;
+
+  private Repository proxyRepo;
+
+  private Server server;
+
+  @Configuration
+  public static Option[] configureNexus() {
+    return NexusPaxExamSupport.options(
+        NexusITSupport.configureNexusBase(),
+        nexusFeature("org.sonatype.nexus.plugins", "nexus-repository-p2")
+    );
+  }
+
+  @Before
+  public void setup() throws Exception {
+    server = Server.withPort(0)
+        .serve("/" + VALID_PACKAGE_URL)
+        .withBehaviours(Behaviours.file(testData.resolveFile(PACKAGE_NAME)))
+        .serve("/" + ARTIFACT_JAR)
+        .withBehaviours(Behaviours.file(testData.resolveFile(ARTIFACT_JAR)))
+        .serve("/" + ARTIFACT_XML_TEST_PATH)
+        .withBehaviours(Behaviours.file(testData.resolveFile(ARTIFACT_XML)))
+        .serve("/" + ARTIFACT_XML_XZ_TEST_PATH)
+        .withBehaviours(Behaviours.file(testData.resolveFile(ARTIFACT_XML_XZ)))
+        .serve("/" + P2_INDEX)
+        .withBehaviours(Behaviours.file(testData.resolveFile(P2_INDEX)))
+        .serve("/folder/" + P2_INDEX)
+        .withBehaviours(Behaviours.file(testData.resolveFile(P2_INDEX)))
+        .serve("/" + COMPOSITE_ARTIFACTS_JAR)
+        .withBehaviours(Behaviours.file(testData.resolveFile(COMPOSITE_ARTIFACTS_JAR)))
+        .start();
+
+    proxyRepo = repos.createP2Proxy("p2-test-proxy", server.getUrl().toExternalForm());
+    proxyClient = p2Client(proxyRepo);
+  }
+
+  @Test
+  public void unresponsiveRemoteProduces404() throws Exception {
+    assertThat(status(proxyClient.get(BAD_PATH)), is(HttpStatus.NOT_FOUND));
+  }
+
+  @Test
+  public void retrieveJarFromProxyWhenRemoteOnline() throws Exception {
+    assertThat(status(proxyClient.get(VALID_PACKAGE_URL)), is(HttpStatus.OK));
+
+    final Asset asset = findAsset(proxyRepo, "/" + VALID_PACKAGE_URL);
+    assertThat(asset.name(), is(equalTo("/" + VALID_PACKAGE_URL)));
+    assertThat(asset.contentType(), is(equalTo(MIME_TYPE)));
+    assertThat(asset.format(), is(equalTo(FORMAT_NAME)));
+
+    final Component component = findComponent(proxyRepo, COMPONENT_NAME);
+    assertThat(component.version(), is(equalTo(VERSION_NUMBER)));
+    assertThat(component.group(), is(equalTo(null)));
+  }
+
+  @Test
+  public void retrieveP2IndexFromProxyWhenRemoteOnline() throws Exception {
+    assertThat(status(proxyClient.get(P2_INDEX)), is(HttpStatus.OK));
+
+    final Asset asset = findAsset(proxyRepo, P2_INDEX);
+    assertThat(asset.name(), is(equalTo(P2_INDEX)));
+    assertThat(asset.format(), is(equalTo(FORMAT_NAME)));
+    // TODO: We discovered these files do exist, but right now it's not working, we need to fix this!
+    // assertThat(status(proxyClient.get("folder/" + P2_INDEX)), is(HttpStatus.OK));
+  }
+
+  @Test
+  public void retrieveCompositeArtifactsJarFromProxyWhenRemoteOnline() throws Exception {
```

Comment:
@aornatovskyy we added these so we could do some refactoring in another PR, we discovered a bug that is fairly nasty and decided these tests would help us figure it out quicker